Allow derived settings to replace previously-defined but non-default settings

dansanduleac opened the pull request:
Following #1034. It remains to update all uses of `DerivedSetting` internally so that they still get created as `DefaultSetting`s, by doing so explicitly.

Commit:
DerivedSetting not a DefaultSetting anymore

Commit:
Derived settings to replace their DerivedSetting, not go at the beginning

harrah commented:
(Note to potential reviewers: as the PR description indicates, this still needs existing uses of derived to be made explicit default settings before review and test.)

Commit:
Derive settings only under the scope of the DerivedSetting

Commit:
Decouple DefaultSetting from Setting/DerivedSetting; BuildCommon.derive() produces default settings by default

dansanduleac commented:
Added the changes mentioned in the issue (first commit), as well as making the derived settings in Defaults be `DefaultSetting`s.

Commit:
Put sbt's derived settings under GlobalScope

dansanduleac commented:
And now actually fixed sbt's derived usages as well, they _have_ to be explicitly Global because otherwise they will only be scoped in `ThisScope` (and more specific scopes), and that causes unresolved reference errors because some dependers expect it to exist in Global as well.
Should be good to go now.

Commit:
Optimise scope intersection for GlobalScope

harrah commented:
Ok, looks good except for a couple minor notes.  Also, tests would be good.

Commit:
Couple of fixes

dansanduleac commented:
Ok, done those. Working on tests (might not get them in til new year)

jsuereth commented:
hey @dansanduleac do you have a chance to talk about this (and other fixes) over skype?  Send me an email or ping me on google +.

Thanks!
- Josh

dansanduleac commented:
@jsuereth yep see email

dansanduleac commented:
Sorry, I forgot about this PR for the past few months, because I didn't explicitly need it anymore, and other things had caught my immediate attention.

dansanduleac committed:
Improve SettingsExample to allow orthogonal scopes (like projects/tasks) at a certain nestIndex

dansanduleac committed:
Add 2 derived settings tests:

1) non-default derived settings, if they produce anything, the settings
   they produce must supersede previous assignents (in the settings seq)
   to the same key.

2) even if a derived setting is scoped at a higher scope (e.g.
   ThisBuild) the settings it produces are scoped at the intersection of
   that (the defining) scope and the scope of the triggering dependency.

2 is particularly nice as it enables this behaviour:
    derive(b in ThisBuild := a.value + 1)
    a in project1 := 0
    // a could be defined in all projects
    ==>
    Now (b in project1).value == (a in project1).value + 1 == 1
    and similarly in all other projects
    all with a single derived setting

dansanduleac commented:
This seems to work, the unit tests pass on my machine.
One thing to note is that at the moment, you still can't use a derived setting to _update_ a setting (i.e. a setting that depends on itself). That might not be very difficult to fix, but atm the derived setting just won't trigger if the only dependency is the setting itself, because that dependency won't  be returned by `Setting.dependencies`.

jsuereth commented:
Yeah, the timeouts we only need to worry about shuffling tests to keep travis happy.

Thanks for adding the tests. will look at this for the 0.13 (0.13.6 now) branch shortly.  I'm currently merging 0.13.2 doc fixes into 0.13, then once 0.13.5 is forked off, we can review/merge this.

jsuereth commented:
Hey @dansanduleac it looks like this broke the `project/derived` scripted test.  I may have to back out the commits for the time being.

Specifically, it looks like the derived settings in the test aren't actually getting used at all.

dansanduleac commented:
Oops that's odd. I'll take a look.

On Wed, May 7, 2014 at 1:01 PM, Josh Suereth notifications@github.com
wrote:

> Hey @dansanduleac it looks like this broke the `project/derived` scripted test.  I may have to back out the commits for the time being.
> 
> ## Specifically, it looks like the derived settings in the test aren't actually getting used at all.
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/sbt/sbt/pull/1036#issuecomment-42418373
